### PR TITLE
Move `GenPollCommand`'s default Button1 callback to `ThreadPoolText` class.

### PR DESCRIPTION
**Pull request by mrtnvgr**
This PR adds default `force_update` callback to `GenPollUrl` class just like in the `GenPollCommand` class.

### CODE REVIEW

**Comment by elParaguayo** [comment]
I'm not sure this needs to be hard coded in the widget. You could just do this in your config by setting `mouse_callbacks={"Button1": lazy.widget["openweather"].force_update()}`

**Comment by mrtnvgr** [comment]
> I'm not sure this needs to be hard coded in the widget. You could just do this in your config by setting `mouse_callbacks={"Button1": lazy.widget["openweather"].force_update()}`

I think it might be useful not only for me, but in any case you decide. Other widgets do similar approach (e.g. `do_execute` in `CheckUpdates` widget)

**Comment by mrtnvgr** [comment]
I noticed that `force_update` callback sets to every widget based on `GenPollCommand`, maybe we do this with `GenPollUrl` too?

File: libqtile/widget/generic_poll_text.py [hunk 1]
```diff
@@ -61,6 +61,10 @@ def __init__(self, **config):
         if self.data and not isinstance(self.data, str):
             self.data = json.dumps(self.data).encode()
 
+    def _configure(self, qtile, bar):
+        base.ThreadPoolText._configure(self, qtile, bar)
+        self.add_callbacks({"Button1": self.force_update})
```
Comment (elParaguayo) [inline libqtile/widget/generic_poll_text.py:66]: This should not be done in `_configure`. Widgets typically set their callbacks during `__init__` so, if we change it here, any custom settings in widgets will be overriden when the widget is configured.

File: libqtile/widget/generic_poll_text.py [hunk 2]
```diff
@@ -61,6 +61,10 @@ def __init__(self, **config):
         if self.data and not isinstance(self.data, str):
             self.data = json.dumps(self.data).encode()
 
+    def _configure(self, qtile, bar):
+        base.ThreadPoolText._configure(self, qtile, bar)
```
Comment (elParaguayo) [inline libqtile/widget/generic_poll_text.py:65]: Why not just do this in `ThreadPoolText`? That's the class that defines the `force_update` method.

**Review by elParaguayo — CHANGES_REQUESTED** [review]
I've added a couple of comments but, to be honest, I'm still not 100% convinced of the need for this.

File: libqtile/widget/generic_poll_text.py (hunk 1 above)
Comment (mrtnvgr) [reply libqtile/widget/generic_poll_text.py:66]: Oh, but `GenPollCommand` does it in `_configure`

**Comment by mrtnvgr** [comment]
> I've added a couple of comments but, to be honest, I'm still not 100% convinced of the need for this.

With this and #4085, user can make a widget which updates only on mouse presses.

**Comment by elParaguayo** [comment]
> > I've added a couple of comments but, to be honest, I'm still not 100% convinced of the need for this.
> 
> With this and #4085, user can make a widget which updates only on mouse presses.

It doesn't _need_ this PR (the button configuration can be done in users' own config) but it's a convenience.

**Comment by mrtnvgr** [comment]
> > > I've added a couple of comments but, to be honest, I'm still not 100% convinced of the need for this.
> > 
> > With this and #4085, user can make a widget which updates only on mouse presses.
> 
> It doesn't _need_ this PR (the button configuration can be done in users' own config) but it's a convenience.

Isn't convenience a good thing?

**Comment by mrtnvgr** [comment]
> This should not be done in `_configure`. Widgets typically set their callbacks during `__init__` so, if we change it here, any custom settings in widgets will be overriden when the widget is configured.

 I do this in `__init__`, some tests fail.

**Comment by elParaguayo** [comment]
> > > > I've added a couple of comments but, to be honest, I'm still not 100% convinced of the need for this.
> > > 
> > > 
> > > With this and #4085, user can make a widget which updates only on mouse presses.
> > 
> > 
> > It doesn't _need_ this PR (the button configuration can be done in users' own config) but it's a convenience.
> 
> Isn't convenience a good thing?

Potentially. But not if it causes unexpected behaviour for users. Some users may not want their widgets updated when they click on them but this change forces that situation on them. To restore their normal behaviour, they now need to make extra config changes. That's not "convenient" for them.

This isn't solving an issue that impacts everybody. That's why I'm hesitant here.

**Comment by elParaguayo** [comment]
> > This should not be done in `_configure`. Widgets typically set their callbacks during `__init__` so, if we change it here, any custom settings in widgets will be overriden when the widget is configured.
> 
> If I do this in `__init__`, some tests fail.

Understanding _why_ they fail would be useful.

**Comment by m-col** [comment]
+1 for not applying this at the level of `ThreadPoolText` to not change things for most existing configs. The reality of qtile is that some things the perfectionist programmer would want to change have to be lived with as users prefer stability over consistency or convenience :laughing:

That said, it does seem intuitive to me for this to be the default for the generic polling widgets in `generic_poll_text.py`, though it's only the case for one of them, and that fact is not documented and one would only discover this through testing or reading the code. To avoid the disruption perhaps it would be best just to add a line to the docstrings of the widgets saying how they can achieve this (i.e. @elParaguayo's first comment), and for `GenPollCommand` that it will update when clicked.

**Pull request closed without merging**